backend/connect.ts

```diff
@@ -0,0 +1,81 @@
+import { IncomingMessage } from "http";
+import { cookieSchema } from "../schemas/poke";
+import { parse } from "url";
+import { WebSocket } from "ws";
+import { getClientRecord, setClientRecord } from "./data";
+import { transact } from "./db";
+import { dispatchMessage } from "./dispatch-message";
+import { Client, ClientMap } from "./server";
+import { Loop } from "./loop";
+
+export async function connect(
+  ws: WebSocket,
+  req: IncomingMessage,
+  clients: ClientMap,
+  loop: Loop
+) {
+  const url = parse(req.url!, true);
+  const parts = (url.pathname ?? "").split("/");
+
+  if (parts[1] != "d" || !parts[2]) {
+    ws.send("invalid url - no room id");
+    ws.close();
+    return;
+  }
+
+  const roomID = parts[2];
+  const clientID = (url.query.clientID as string | undefined) ?? "";
+  if (!clientID) {
+    ws.send("invalid url - no client ID");
+    ws.close();
+    return;
+  }
+  const baseCookie = cookieSchema.safeParse(
+    JSON.parse((url.query.baseCookie as string | undefined) ?? "null")
+  );
+  if (!baseCookie.success) {
+    ws.send(`invalid base cookie in url: ${baseCookie.error.issues}`);
+    ws.close();
+    return;
+  }
+
+  const existingClient = clients.get(clientID);
+  if (existingClient) {
+    existingClient.socket.close();
```

Comment:
> Uh oh it's possible that the async close from 2 above completes after step 3 above, which will delete the newly added client from the clients map.

Whoops, yeah. I didn't unit test connect yet, and likely would have noticed this then. An easy fix is to ensure that the instance of the client removed matches (by identity, not client ID) the one the event fired upon.

> I think for sanity it might be a good idea to wait for some operations like close to complete

Perhaps.